components/claim/components/email_to_ldap.tsx

```diff
@@ -11,81 +10,95 @@ import {t} from 'utils/i18n.jsx';
 import LoginMfa from 'components/login/login_mfa.jsx';
 import LocalizedInput from 'components/localized_input/localized_input';
 
-export default class EmailToLDAP extends React.PureComponent {
-    static propTypes = {
-        email: PropTypes.string,
-        siteName: PropTypes.string,
-        ldapLoginFieldName: PropTypes.string,
-    };
+type Props = {
+    email?: any,
+    siteName?: any,
+    ldapLoginFieldName?: any,
+};
 
-    constructor(props) {
-        super(props);
 
-        this.state = {
-            passwordError: '',
-            ldapError: '',
-            ldapPasswordError: '',
-            serverError: '',
-            showMfa: false,
-        };
+type State = {
+    passwordError?: string,
+    ldapError?: string,
+    ldapPasswordError?: string,
+    serverError?: string,
+    showMfa?: boolean,
+};
+
+export default class EmailToLDAP extends React.PureComponent <Props, State>{
+    private emailPasswordInput: React.RefObject<HTMLInputElement>;
+    private ldapIdInput: React.RefObject<HTMLInputElement>;
+    private ldapPasswordInput: React.RefObject<HTMLInputElement>;
+
+    public constructor(props: Props) {
+        super(props);
 
         this.emailPasswordInput = React.createRef();
         this.ldapIdInput = React.createRef();
         this.ldapPasswordInput = React.createRef();
+        this.state = this.getDefaultState();
     }
 
-    preSubmit = (e) => {
-        e.preventDefault();
-
-        var state = {
+    getDefaultState() {
```

Comment:
this should return type state, doesn't it?

Reply:
I don't quite understand this I'm sorry

Reply:
by not definining the return type, it might be any, so you are returning some object with properties that might or not match what it is expected by the caller. having a type for it will ensure that the compiler can actually help you figuring out if you forgot to add something in the code.

Reply:
So I'm returning type state like this?

`getDefaultState() {
        type State = {
            passwordError: '',
            ldapError: '',
            ldapPasswordError: '',
            serverError: '',
            ldapId: '',
            password: '',
            ldapPassword: '',
            showMfa: false
        };
    }`

Reply:
If above is correct, how do I access the state fields?
![Screen Shot 2021-06-23 at 12 51 09](https://user-images.githubusercontent.com/35423212/123037833-92607d00-d419-11eb-993b-df5d53f9122c.png)

Reply:
it doesn't change how you access them, it just tells the compiler "this is the object I'll return, check that I didn't forget to add any properties or added any unexpected ones"

Reply:
Ok so type state as I capture above is correct?

Reply:
yes, but you have to use it for that error to go away

Reply:
so how do I use it for the error to go away? I'm confuse about how to get State object inside the function

Reply:
You should define it outside of the function and add it as the return type.
```ts
type StateType = {...}

getDefaultState(): StateType {
return ....
}
```